### ISSUE TITLE
Sync gets stuck and must be terminated/restarted manually in order to work

### ISSUE THREAD

**Issue by marcozov** [argoproj/argo-cd#4811]
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

I'm trying to deploy an ArgoCD Application that contains a configmap, two certificates (Certificate custom resource from cert-manager) and a KafkaConnect instance (from Strimzi operator).

I defined the following annotations: argocd.argoproj.io/sync-wave (to make sure to have configmap and certificates before the kafkaconnect instance) and argocd.argoproj.io/sync-options on CRDs. When the application is deployed, the sync gets stuck: it keeps saying OutOfSync and Syncing (see attached image). However, if I stop the sync (click on Syncing, terminate) and then Sync the application again, then it successfully deploys all the defined resources.

Although I am using Custom Resources here (Certificate from cert-manager and KafkaConnect from Strimzi), the related custom health seem to exist already (https://github.com/argoproj/argo-cd/tree/master/resource_customizations).

The main problem is that I have several applications of this kind, so I would like to be able to automate this (instead of relying on manually stopping the sync and restarting it for all these applications). Any idea?

**To Reproduce**

Deploy an ArgoCD Application that contains the resources mentioned above. The Sync phase will start by itself and get stuck.

**Expected behavior**

The Sync should not get stuck and continue, without needing any manual action (terminate Sync and start it again)

**Screenshots**

<img width="1797" alt="argocd" src="https://user-images.githubusercontent.com/17951561/98747024-f7012700-23b6-11eb-9445-5bc28463cede.png">

**Version**

```shell
argocd: v1.7.6+b04c25e
  BuildDate: 2020-09-19T00:50:44Z
  GitCommit: b04c25eca8f1660359e325acd4be5338719e59a0
  GitTreeState: clean
  GoVersion: go1.14.1
  Compiler: gc
  Platform: linux/amd64
argocd-server: v1.7.6+b04c25e
  BuildDate: 2020-09-19T00:52:04Z
  GitCommit: b04c25eca8f1660359e325acd4be5338719e59a0
  GitTreeState: clean
  GoVersion: go1.14.1
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: {Version:kustomize/v3.6.1 GitCommit:c97fa946d576eb6ed559f17f2ac43b3b5a8d5dbd BuildDate:2020-05-27T20:47:35Z GoOs:linux GoArch:amd64}
  Helm Version: version.BuildInfo{Version:"v3.2.0", GitCommit:"e11b7ce3b12db2941e90399e874513fbd24bcb71", GitTreeState:"clean", GoVersion:"go1.13.10"}
  Kubectl Version: v1.17.8
```

**Comment by Bemosi** [comment]
I have been playing with a similar issue on version `v1.7.8+ef5010c`, in an app of apps scenario.

When the syncing hit a "Degraded" state, as part of the certificate issuing, it seems that the application syncing started waiting on everything all over again and would never get the healthy notifications.

I think it is an issue with the certificate health check, I overrode the default check by changing the `argocd-cm.yaml` file with the below (only replacing `Degraded` with `Progressing`):
```
data:
  resource.customizations: |
    cert-manager.io/Certificate:
      health.lua: |
        hs = {}
        if obj.status ~= nil then
          if obj.status.conditions ~= nil then
            for i, condition in ipairs(obj.status.conditions) do
              if condition.type == "Ready" and condition.status == "False" then
                hs.status = "Progressing"
                hs.message = condition.message
                return hs
              end
              if condition.type == "Ready" and condition.status == "True" then
                hs.status = "Healthy"
                hs.message = condition.message
                return hs
              end
            end
          end
        end

        hs.status = "Progressing"
        hs.message = "Waiting for certificate"
        return hs
```

Seems to have resolved the issue for me (at least in the very small number of tests I have done since).

**Comment by marcozov** [comment]
Thanks a lot, this seems to work for this specific ArgoCD Application! On other applications I have similar issues that require further investigation, but now at least I know how I can start approaching the problem.

Just wondering: it looks like the Certificate health check proposed by argocd itself (mentioned in https://argoproj.github.io/argo-cd/operator-manual/health/ and in https://github.com/argoproj/argo-cd/tree/master/resource_customizations/cert-manager.io/Certificate) causes this issue. Do you think it would make sense to open a PR to change that check, in order to fix that?

**Comment by cmaster11** [comment]
I can say this is happening also to us.

We have a very odd situation, where we have 4 environments, with pretty much the same configuration, e.g. same ArgoCD application deployed in all of them. And only ONE of those, very frequently has this sync issue.

I'd like to understand how to debug it, because there is no apparent problem, other than random syncs getting stuck D:

Our setup does not include custom health checks, but we do have sync waves.

**Comment by jessesuen** [comment]
> Just wondering: it looks like the Certificate health check proposed by argocd itself (mentioned in https://argoproj.github.io/argo-cd/operator-manual/health/ and in https://github.com/argoproj/argo-cd/tree/master/resource_customizations/cert-manager.io/Certificate) causes this issue. Do you think it would make sense to open a PR to change that check, in order to fix that?

Yes, if the health check is not functioning, please send a PR.

**Comment by Bemosi** [comment]
Thinking about this a bit more - It may be that the application syncing issue is just a symptom of a wider issue?
The health check does eventually get to a Healthy state and is in a "Degraded" state when the certificate issuance is pending.

I can make a PR for the small change above, which reduces issues when using cert-manager certificates, but that won't fix the underlying sync issue. Where after getting into a degraded state, the application will wait for all resources to report "healthy" and seemingly deadlocks or waits for an event that never arrives.

**Comment by lifelofranco** [comment]
I would just like to add that this also happens on the example in the docs, about configuring a slack notification at postSync.

<img width="459" alt="Screen Shot 2021-01-19 at 12 44 22 PM" src="https://user-images.githubusercontent.com/4952784/104988912-3f426280-5a54-11eb-9ba8-a6a8ba336cfa.png">

ArgoCD gets stuck at syncing, but terminating it manually and syncing it again would always fix the issue and send the slack notif.

**Comment by debu99** [comment]
i have the same issue as @lifelofranco mentioned
Argo CD
v1.8.3+0f9c684
Build Date
2021-01-21T22:20:39Z

**Comment by lifelofranco** [comment]
@rbreeze, I was running v1.7.6, the default appVersion of the helm chart at the time. Also encountered issues where deletion gets stuck.

However, I have since updated to v1.8.3 and I am no longer facing any issues! Everything working as expected.

**Comment by sathish-kumar-narayanan** [comment]
I had a similar problem. I disabled the auto sync from app, and did explicit app sync and it solved the issue for me.

**Comment by ishitasequeira** [comment]
I would like to work on this issue.

**Comment by sanzenwin** [comment]
I'm facing this issues too.
ArgoCD v2.0.1+33eaf11

**Comment by Kampe** [comment]
Seeing the same issues with 2.0.3

**Comment by Kampe** [comment]
it's possible given invalid k8s manifest yaml to get Argocd to refuse to terminate jobs with 2.0.5, very easy to accomplish.

**Comment by gajus** [comment]
@alexmt Has this been addressed in v2.3?

**Comment by pre** [comment]
We're seeing this issue in v2.3.1. We have a `Job` which executes with 
```
  annotations:
    argocd.argoproj.io/hook: Sync
    argocd.argoproj.io/hook-delete-policy: HookSucceeded,BeforeHookCreation
```

We've tried with `PreSync` and only `HookSucceeded` and only `BeforeHookCreation` but the effect is always the same.

This works 98% in cases. But 2% of deployments get stuck with "waiting for job-name-here".

When the Sync is manually terminated, the next Sync will succeed 99% of time.

We're having such a number of deployments that we see this error at least once every day 🤔 
Updating to ArgoCD >2.3 is on our backlog but requires work due to changes in ConfigManagementPlugin. 
However, I didn't see anything in ArgoCD release notes up-to 2.5.x which would touch this particular problem.

**Comment by pre** [comment]
The "waiting for completion of hook xx" stays rolling for an indefinite period of time. 

The only way is to unblock ArgoCD is to manually terminate the sync and restart it. Eg deleting an App via ArgoCD HTTP api gets into "queue" but is not processed, because "sync is in progress".

<img width="593" alt="image" src="https://user-images.githubusercontent.com/63895/208426523-47b21809-2e38-4eda-9ebd-30bc8a680a4d.png">

**Comment by pre** [comment]
One interesting symptom: When ArgoCD is stuck with "waiting for completion of hook ..." no operation works with that Application resource until the sync operation is manually terminated. 

For example: 
* Application gets stuck with "waiting for completion of hook .." 
* Application received DELETE via the ArgoCD HTTP API
* Nothing changes, Application can stays in this state for days until manually terminated.
* Once the sync is manually terminated, the deletion (issued days ago) immediately proceeds and Application is deleted.

**Comment by paolostancato** [comment]
I'm experiencing something similar, in my case the workflow is like this:

we use the PR Generator to create dev environments, so we create a branch, let's call it `dev/branch`, then the PR Generator deploys an ApplicationSet with a bunch of Argo ApplicationSet

If I close the PR (and delete the branch), the Argo apps that were syncing would end up with this error:

```
ComparisonError rpc error: code = Unknown desc = Unable to resolve 'dev/branch' to a commit SHA
```

forever syncing, until someone terminates the sync process where the app would, in fact, terminates.

I would expect the app to be killed if there's no branch standing for it.

**Comment by PetrPhilipp** [comment]
Hi, Im having a very similar issue, is there at least a way to automatically terminate a long running sync? I could not find a setting like this.

**Comment by blakepettersson** [comment]
See #16630
